File: graph/cycle_detection/py/dfs.py
WHITE = 0
GREY = 1
BLACK = 2

def has_cycle(g):
    color = {}
    for vertex in g:
        color[vertex] = WHITE
    for vertex in g:
        if color[vertex] == WHITE:
            if dfs(g, vertex, color):
                return True
    return False

def dfs(g, vertex, color):
    color[vertex] = GREY
    for u in g[vertex]:
        if color[u] == GREY:
            return True
        if color[u] == WHITE:
            if dfs(g, u, color):
                return True
    color[vertex] = BLACK
    return False

File: graph/cycle_detection/java/DFS.java
public class DFS {
  public static boolean hasCycle(int[][] graph) {
    int n = graph.length;
    int[] color = new int[n];
    for (int u = 0; u < n; u++) {
      if (color[u] == WHITE) {
        if (explore(graph, u, color)) {
          return true;
        }
      }
    }
    return false;
  }

  private static boolean explore(int[][] graph, int u, int[] color) {
    color[u] = GREY;
    for (int v : graph[u]) {
      if (color[v] == GREY) return true;
      if (color[v] == WHITE) {
        if (explore(graph, v, color)) {
          return true;
        }
      }
    }
    color[u] = BLACK;
    return false;
  }

  private static final int WHITE = 0;
  private static final int GREY = 1;
  private static final int BLACK = 2;
}
